### PR TITLE
ML.NET image classification with TensorFlow tutorial

### LINKED ISSUE #5379: Image recognition sample for ML.NET
I'd love to see a sample that trains and recognizes images (ie object detection)

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: e855e630-1a30-d138-4ac6-bfeefebf9aae
* Version Independent ID: afc0a7d1-d3c2-acf2-d4ea-d2565e22b237
* Content: [ML.NET tutorials](https://docs.microsoft.com/en-us/dotnet/machine-learning/tutorials/)
* Content Source: [docs/machine-learning/tutorials/index.md](https://github.com/dotnet/docs/blob/master/docs/machine-learning/tutorials/index.md)
* Service: **dotnet-ml**
* GitHub Login: @JRAlexander
* Microsoft Alias: **johalex**

### PR DESCRIPTION
**Pull request by JRAlexander**
TensorFlow Transfer Learning for Image Classification Tutorial 

Fixes #5379 

[Internal Review URL](https://review.docs.microsoft.com/en-us/dotnet/machine-learning/tutorials/image-classification?branch=pr-en-us-11383) 
[Related Code Sample](https://github.com/dotnet/samples/pull/727)

[AB#1450815](https://dev.azure.com/mseng/web/wi.aspx?pcguid=0efb4611-d565-4cd1-9a64-7d6cb6d7d5f0&id=1450815)

### CODE REVIEW

File: docs/machine-learning/tutorials/tensorflow-transfer-learning.md [hunk 1]
```diff
@@ -0,0 +1,337 @@
+---
+title: Use ML.NET in a TensorFlow transfer learning scenario for image classification
```
Comment (asthana86) [inline docs/machine-learning/tutorials/tensorflow-transfer-learning.md]: It's very unlikely our user base would understand what transfer learning scenario means and what value does that bring in. Leading from the scenario perspective here would help, rewording this title to "Creating a custom Image Classifier using TensorFlow and ML.NET" or "Re-train an existing TensorFlow Image Classifier with ML.NET to add new categories" would be a better title.

File: docs/machine-learning/tutorials/tensorflow-transfer-learning.md [hunk 2]
```diff
@@ -0,0 +1,337 @@
...
+> * Understand the problem
+> * Build and Train the model
+> * Predict with the trained model
+> * Deploy and Predict with a loaded model
+
+You can find the source code for this tutorial at the [dotnet/samples](https://github.com/dotnet/samples/tree/master/machine-learning/tutorials/TransferLearningTF) repository.
+
+## Understand the problem
+
+Image classification is a common Machine Learning problem which has well-known solutions. This tutorial uses an approach that mixes new techniques (deep learning) and traditional (Logistic Regression) techniques.
```
Comment (asthana86) [inline docs/machine-learning/tutorials/tensorflow-transfer-learning.md]: At a high level also talk about what Deep Learning is and why we are mixing Deep Learning with Traditional machine learning techniques here?

File: docs/machine-learning/tutorials/tensorflow-transfer-learning.md [hunk 3]
```diff
@@ -0,0 +1,337 @@
...
+
+This model uses the [Inception model](https://storage.googleapis.com/download.tensorflow.org/models/inception5h.zip) as a *featurizer* (the model is already stored in the [assets folder](./ImageClassification.Train/assets/inputs/inception/)). This means that the model processes input images through the neural network, using the tensor output before classification. This tensor contains the *image features*, used for image identification.
```
Comment (asthana86) [inline docs/machine-learning/tutorials/tensorflow-transfer-learning.md]: Provide some detail on what the pre-trained model InceptionV3 is. What is the default setup of InceptionV3, different categories it supports and how and why are we trying to customize it.

Comment (asthana86) [inline docs/machine-learning/tutorials/tensorflow-transfer-learning.md]: Its very hard to follow what these lines are trying to explain. For the audience we are targeting, can you help explain what a image featurizer is? What is a tensor? What is a tensor output?

File: docs/machine-learning/tutorials/tensorflow-transfer-learning.md [hunk 4]
```diff
@@ -0,0 +1,337 @@
...
+
+Finally, these image features will be fed to a Logistic regression algorithm which will learn how to classify different sets of image features.
```
Comment (asthana86) [inline docs/machine-learning/tutorials/tensorflow-transfer-learning.md]: Why did we use a logistic regression algorithm and what did we customize and why?

File: docs/machine-learning/tutorials/tensorflow-transfer-learning.md [hunk 5]
```diff
@@ -0,0 +1,337 @@
...
+
+## DataSet
+
+There are two data sources: the `tsv` file and the image files.  The tags.tsv file contains two columns: the first one is defined as `ImagePath` and the second one is the `Label` corresponding to the image. The following example file does not have a header row, and looks like this:
```
Comment (asthana86) [inline docs/machine-learning/tutorials/tensorflow-transfer-learning.md]: Please explain why we need this dataset given we are starting of with a pre-trained model? What are we trying to achieve with this data.

File: docs/machine-learning/tutorials/tensorflow-transfer-learning.md [hunk 6]
```diff
@@ -0,0 +1,337 @@
...
+
+The training and testing images are located in the assets folders. These images belong to Wikimedia Commons.
+> *[Wikimedia Commons](https://commons.wikimedia.org/w/index.php?title=Main_Page&oldid=313158208), the free media repository.* Retrieved 10:48, October 17, 2018 from:  
+> https://commons.wikimedia.org/wiki/Pizza  
+> https://commons.wikimedia.org/wiki/Toaster  
+> https://commons.wikimedia.org/wiki/Teddy_bear  
+
+## ML Task - [Image Classification](https://en.wikipedia.org/wiki/Outline_of_object_recognition)
+
+To solve this problem, first you'll build an ML model. Then, train the model on existing data, evaluate the model quality, and consume the model to classify a new image.
```
Comment (asthana86) [inline docs/machine-learning/tutorials/tensorflow-transfer-learning.md]: Model already exists, we are merely retraining the model with a fraction of the dataset.

File: docs/machine-learning/tutorials/tensorflow-transfer-learning.md [hunk 7]
```diff
@@ -0,0 +1,337 @@
...
+
+4. Install the **Microsoft.ML NuGet Package**:
+
+    In Solution Explorer, right-click on your project and select **Manage NuGet Packages**. Choose "nuget.org" as the Package source, select the Browse tab, search for **Microsoft.ML**. Click on the **Version** drop down, select the **0.10.0** package in the list, and select the **Install** button. Select the **OK** button on the **Preview Changes** dialog and then select the **I Accept** button on the **License Acceptance** dialog if you agree with the license terms for the packages listed. Repeat these steps for **Microsoft.ML.ImageAnalytics v0.10.0** and **Microsoft.ML.TensorFlow v0.10.0**.
+
+  > [!NOTE]
+  > This tutorial uses **Microsoft.ML v0.10.0**, **Microsoft.ML.ImageAnalytics v0.10.0**, and **Microsoft.ML.TensorFlow v0.10.0**.
+
+### Create classes and define paths
+
```
Comment (asthana86) [inline docs/machine-learning/tutorials/tensorflow-transfer-learning.md]: Provide link to the sample code upfront, its very hard to follow these singular steps of copy pasting code. It's also error prone.

File: docs/machine-learning/tutorials/tensorflow-transfer-learning.md [hunk 8]
```diff
@@ -0,0 +1,337 @@
...
+Remove the existing class definition and add the following code, which has two classes `ImageNetData` and `ImageNetPipeline`, to the *ImageNetData.cs* file:
+
+[!code-csharp[DeclareGlobalVariables](../../../samples/machine-learning/tutorials/TransferLearningTF/ImageNetData.cs#DeclareTypes)]
+
+`ImageNetData` is the input image data class and has the following <xref:System.String> fields:
+
+* `ImagePath` contains the path for the image file
+* `Label` contains a value for the image label
+
+`ImageNetData` also has a method (`ReadFromCsv`) to read the image data `.csv` file and load it into an `ImageNetData` object.
```
Comment (asthana86) [inline docs/machine-learning/tutorials/tensorflow-transfer-learning.md]: Why does the ImageNetData class have a ReadFromCSV Method?

File: docs/machine-learning/tutorials/tensorflow-transfer-learning.md [hunk 9]
```diff
@@ -0,0 +1,337 @@
...
+
+`ImageNetPipeline` is the class used for inspecting column values after the model has been trained. It has a `string` (`ImagePath`) for the image path. The `Label` is used to create and train the model, and it's also used to evaluate the model. The `PredictedLabelValue` is used during prediction and evaluation. For evaluation, an input with training data, the predicted values, and the model are used.
```
Comment (asthana86) [inline docs/machine-learning/tutorials/tensorflow-transfer-learning.md]: Hard to follow. 
What is a column value? 
How is the label being used to create and train a model and also for evaluation in the same class?

File: docs/machine-learning/tutorials/tensorflow-transfer-learning.md [hunk 10]
```diff
@@ -0,0 +1,337 @@
...
+Remove the existing class definition and add the following code, which has the `ImageNetPrediction` class, to the *ImageNetPrediction.cs* file:
+
+[!code-csharp[DeclareGlobalVariables](../../../samples/machine-learning/tutorials/TransferLearningTF/ImageNetPrediction.cs#DeclareTypes)]
+
+`ImageNetPrediction` is the image prediction class and has the following fields:
+
+* `Score` contains the confidence percentage for a given image classification.
+* `PredictedLabelValue` contains a value for the predicted image classification label.
+
+`IssuePrediction` is the class used for prediction after the model has been trained. It has a `string` (`ImagePath`) for the image path. The `Label` is used to create and train the model, and it's also used with a second dataset to evaluate the model. The `PredictedLabelValue` is used during prediction and evaluation. For evaluation, an input with training data, the predicted values, and the model are used.
```
Comment (asthana86) [inline docs/machine-learning/tutorials/tensorflow-transfer-learning.md]: What is IssuePrediction?

File: docs/machine-learning/tutorials/tensorflow-transfer-learning.md [hunk 11]
```diff
@@ -0,0 +1,337 @@
...
+
+Initialize the `_mlContext` global variable  with a new instance of `MLContext`.  Replace the `Console.WriteLine("Hello World!")` line with the following code in the `Main` method:
+
+[!code-csharp[CreateMLContext](../../../samples/machine-learning/tutorials/TransferLearningTF/Program.cs#CreateMLContext)]
+
+## Load the data
+
+ As the input and output of [`Transforms`](../basic-concepts-model-training-in-mldotnet.md#transformer), a `DataView` is the fundamental data pipeline type, comparable to `IEnumerable` for `LINQ`.
+
+In ML.NET, data is similar to a `SQL view`. It is lazily evaluated, schematized, and heterogenous. The object is the first part of the pipeline, and loads the data.  The `DataView` is used to create and train the model.
```
Comment (asthana86) [inline docs/machine-learning/tutorials/tensorflow-transfer-learning.md]: "lazily evaluated, schematized, and heterogenous" What does this mean to an average .NET developer? 
DataView is not used to create and train the model. 

Please follow the recommendation tutorial.

File: docs/machine-learning/tutorials/tensorflow-transfer-learning.md [hunk 12]
```diff
@@ -0,0 +1,337 @@
...
+
+Since your previously created `ImageNetData` data model type matches the dataset schema, you can combine the initialization, mapping, and dataset loading into one line of code.
```
Comment (asthana86) [inline docs/machine-learning/tutorials/tensorflow-transfer-learning.md]: what is a data model type?

File: docs/machine-learning/tutorials/tensorflow-transfer-learning.md [hunk 13]
```diff
@@ -0,0 +1,337 @@
...
+Add the following call to the `BuildAndTrainModel`method as the next line of code in the `Main` method:
+
+[!code-csharp[CallBuildAndTrainModel](../../../samples/machine-learning/tutorials/TransferLearningTF/Program.cs#CallBuildAndTrainModel)]
+
+The `BuildAndTrainModel` method executes the following tasks:
+
+* Loads the data
+* Extracts and transforms the data.
+* Scores the TensorFlow model.
+* Trains the model.
```
Comment (asthana86) [inline docs/machine-learning/tutorials/tensorflow-transfer-learning.md]: re-trains the model

File: docs/machine-learning/tutorials/tensorflow-transfer-learning.md [hunk 14]
```diff
@@ -0,0 +1,337 @@
...
+```csharp
+public static void BuildAndTrainModel(MLContext mlContext, string dataLocation, string imagesFolder, string inputModelLocation, string outputModelLocation)
+{
+
+}
+```
+
+## Extract Features and transform the data
+
+Pre-processing and cleaning data are important tasks that occur before a dataset is used effectively for machine learning. Raw data is often noisy and unreliable, and may be missing values. Using data without these modeling tasks can produce misleading results.
```
Comment (asthana86) [inline docs/machine-learning/tutorials/tensorflow-transfer-learning.md]: How does this relate to this particular tutorial?

File: docs/machine-learning/tutorials/tensorflow-transfer-learning.md [hunk 15]
```diff
@@ -0,0 +1,337 @@
...
+
+ML.NET's transform pipelines compose a custom `transforms`set that is applied to your data before training or testing. The transforms' primary purpose is data [featurization](../resources/glossary.md#feature-engineering). Machine learning algorithms understand [featurized](../resources/glossary.md#feature) data, so the next step is to transform our textual data into a format that our ML algorithms recognize. That format is a [numeric vector](../resources/glossary.md#numerical-feature-vector).
```
Comment (asthana86) [inline docs/machine-learning/tutorials/tensorflow-transfer-learning.md]: How does this relate to this particular tutorial, Is there textual data for this tutorial?

File: docs/machine-learning/tutorials/tensorflow-transfer-learning.md [hunk 16]
```diff
@@ -0,0 +1,337 @@
...
+
+When the model is trained and evaluated, by default, the values in the **Label** column are considered as correct values to be predicted. As you are using a pretrained model you need to map fields to the new model.. To do that, use the `MLContext.Transforms.Conversion.MapValueToKey`, which is a wrapper for the <xref:Microsoft.ML.ConversionsExtensionsCatalog.ValueToKeyMappingEstimator%2A> transformation class.  The `MapValueToKey` returns an <xref:Microsoft.ML.Data.EstimatorChain%601> that will effectively be a pipeline. Name this `pipeline` as you will then append the trainer to it. Add the next line of code:
+
+[!code-csharp[MapValueToKey1](../../../samples/machine-learning/tutorials/TransferLearningTF/Program.cs#MapValueToKey1)]
+
+Your image processing pipeline uses pre-trained DNN featurizers for feature extraction. The images are loaded in memory as a Bitmap type. Next, you resize the images as the pre-trained model has a defined input image width and height. Finally, The `ImagePixelExtractingEstimator` extracts the pixels from the input images and converts them into a numeric vector. Add these image transforms as the next lines of code:
```
Comment (asthana86) [inline docs/machine-learning/tutorials/tensorflow-transfer-learning.md]: Explain why we need to re-size images? also what a DNN means here in this context. This is the first mention of DNN in this tutorial.

File: docs/machine-learning/tutorials/tensorflow-transfer-learning.md [hunk 17]
```diff
@@ -0,0 +1,337 @@
...
+
+[!code-csharp[ImageTransforms](../../../samples/machine-learning/tutorials/TransferLearningTF/Program.cs#ImageTransforms)]
+
+The `TensorFlowTransform` extracts specified outputs and scores a dataset using a pre-trained Tensorflow model.
```
Comment (asthana86) [inline docs/machine-learning/tutorials/tensorflow-transfer-learning.md]: Provide more details about what a TensorFlowTransform is. 

Using a picture here showing how we augment a standard ML.NET pipeline with a TFTransform would help here.

File: docs/machine-learning/tutorials/tensorflow-transfer-learning.md [hunk 18]
```diff
@@ -0,0 +1,337 @@
...
+
+[!code-csharp[ImageTransforms](../../../samples/machine-learning/tutorials/TransferLearningTF/Program.cs#ImageTransforms)]
+
+The `TensorFlowTransform` extracts specified outputs and scores a dataset using a pre-trained Tensorflow model. 
+
+[!code-csharp[ScoreTensorFlowModel](../../../samples/machine-learning/tutorials/TransferLearningTF/Program.cs#ScoreTensorFlowModel)]
+
+### Choose a learning algorithm
+
+To add the learning algorithm, call the `mlContext.MulticlassClassification.Trainers.LogisticRegression` wrapper method.  The `LogisticRegression` is appended to the `pipeline` and accepts the featurized `Title` and `Description` (`Features`) and the `Label` input parameters to learn from the historic data. You also need to map the label to the value to return to its original readable state. Do both of those actions with the following code:
```
Comment (asthana86) [inline docs/machine-learning/tutorials/tensorflow-transfer-learning.md]: A few questions here should be answered. 
1. How and why are we trying to add these learners on top of an already pre-trained model? 
2. ML.NET supports a variety of learners for classification, how and why did we pick Logitistic Regression versus others
3. What is a pipeline? (follow concepts in the recommendation tutorial).

File: docs/machine-learning/tutorials/tensorflow-transfer-learning.md [hunk 19]
```diff
@@ -0,0 +1,337 @@
...
+
+[!code-csharp[AddTrainer](../../../samples/machine-learning/tutorials/TransferLearningTF/Program.cs#AddTrainer)]
+
+You also need to map the predictedlabel to the predictedlabelvalue:
+
+[!code-csharp[MapValueToKey2](../../../samples/machine-learning/tutorials/TransferLearningTF/Program.cs#MapValueToKey2)]
+
+You train the model, <xref:Microsoft.ML.Data.TransformerChain%601>, based on the dataset that has been loaded and transformed. Once the estimator has been defined, you train your model using the <xref:Microsoft.ML.Data.EstimatorChain%601.Fit%2A> method while providing the already loaded training data. This returns a model to use for predictions. `pipeline.Fit()` trains the pipeline and returns a `Transformer` based on the `DataView` passed in. This DataView includes the `softmax2_pre_activation` property (also known as image features) and is produced by the ApplyTensorFlowGraph function. The experiment is not executed until the `.Fit()` method runs. Add this with the following code:
```
Comment (asthana86) [inline docs/machine-learning/tutorials/tensorflow-transfer-learning.md]: Please follow the recommendation tutorial. Estimator Chain is not a concept we need to explain, start with explaining transformers and estimators. Estimator Chain is really just the same as an estimator.

Comment (asthana86) [inline docs/machine-learning/tutorials/tensorflow-transfer-learning.md]: What is a DataView?

Comment (asthana86) [inline docs/machine-learning/tutorials/tensorflow-transfer-learning.md]: What is softmax2_pre_activation? Explain this in the context of a scoring function for multi-class classification.

Comment (asthana86) [inline docs/machine-learning/tutorials/tensorflow-transfer-learning.md]: What is ApplyTensorFlowGraph does at a high level?

File: docs/machine-learning/tutorials/tensorflow-transfer-learning.md [hunk 20]
```diff
@@ -0,0 +1,337 @@
...
+
+[!code-csharp[TrainModel](../../../samples/machine-learning/tutorials/TransferLearningTF/Program.cs#TrainModel)]
+
+Notice the use of the `Transform` method of the machine learning `model` (a transformer) to input the features and return predictions.
```
Comment (asthana86) [inline docs/machine-learning/tutorials/tensorflow-transfer-learning.md]: Please simplify. 

"Once the model has been trained, we can use the trained model to make predictions on new data (test data), we do so by using the model.transform API" etc.

File: docs/machine-learning/tutorials/tensorflow-transfer-learning.md [hunk 21]
```diff
@@ -0,0 +1,337 @@
...
+
+
+[!code-csharp[EnumerateModel](../../../samples/machine-learning/tutorials/TransferLearningTF/Program.cs#EnumerateModel)]
+
+
+
+
+[!code-csharp[DisplayInfo](../../../samples/machine-learning/tutorials/TransferLearningTF/Program.cs#DisplayInfo)]
+
+The <xref:Microsoft.ML.MulticlassClassificationCatalog.Evaluate%2A> method computes the quality metrics for the model using the specified dataset. It returns a <xref:Microsoft.ML.Data.MultiClassClassifierMetrics> object that contains the overall metrics computed by multiclass classification evaluators.
```
Comment (asthana86) [inline docs/machine-learning/tutorials/tensorflow-transfer-learning.md]: First time in the tutorial we mention Multi-Classification, this is something that should be presented at the start of the tutorial. 

We are building a customized image classifier using ImageNet. ImageNet as an image classifier allows us to predict between a 1000 categories, since this allows us to classify > 2 categories this is an example of multi-class classifier etc.

**Review by asthana86 — CHANGES_REQUESTED** [review]
At a high level this tutorial needs to be written from a perspective of a .NET developer approaching ML from a scenario perspective. In addition to this let's try to explain in this tutorial how and what are we trying to achieve with Transfer Learning while explaining key concepts along the way. The comments posted here should help in clarifying these further.

File: docs/machine-learning/tutorials/tensorflow-transfer-learning.md [hunk 22]
```diff
@@ -0,0 +1,337 @@
...
+title: Use ML.NET in a TensorFlow transfer learning scenario for image classification
+description: Discover how to use ML.NET in a TensorFlow transfer learning scenario to classify images by reusing a pre-trained TensorFlow model.
+ms.date: 03/20/2019
+ms.topic: tutorial
+ms.custom: mvc
+#Customer intent: As a developer, I want to reuse a pre-trained TensorFlow model with ML.NET so that I can classify images with a small amount of training data.
+---
+# Tutorial: Use ML.NET in a TensorFlow transfer learning scenario for image classification
+
+This sample tutorial illustrates using ML.NET to reuse a pre-trained TensorFlow model to classify images with a small amount of training data via a .NET Core console application using C# in Visual Studio 2017. In the world of machine learning, this is called transfer learning.
```
Comment (asthana86) [inline docs/machine-learning/tutorials/tensorflow-transfer-learning.md]: Please also provide at a high level what style of transfer learning are we trying to achieve with this tutorial. 
Retrain penultimate layer?
Retrain wieghts on different layers?
etc.

File: docs/machine-learning/tutorials/tensorflow-transfer-learning.md [hunk 23]
```diff
@@ -0,0 +1,337 @@
...
+You can find the source code for this tutorial at the [dotnet/samples](https://github.com/dotnet/samples/tree/master/machine-learning/tutorials/TransferLearningTF) repository.
+
+In this tutorial, you learned how to:
+> [!div class="checklist"]
+> * Understand the problem
+> * Build and Train the model
+> * Predict with the trained model
+> * Deploy and Predict with a loaded model
+
+Check out the Machine Learning samples GitHub repository to explore an expanded image classification sample.
```
Comment (asthana86) [inline docs/machine-learning/tutorials/tensorflow-transfer-learning.md]: Next steps here should be trying out different styles of Transfer Learning or other model architectures.

File: docs/machine-learning/tutorials/image-classification.md [hunk 24]
```diff
@@ -0,0 +1,400 @@
...
+[!code-csharp[DeclareGlobalVariables](../../../samples/machine-learning/tutorials/TransferLearningTF/ImageNetData.cs#DeclareTypes)]
+
+`ImageNetData` is the input image data class and has the following <xref:System.String> fields:
+
+* `ImagePath` contains the path for the image file
+* `Label` contains a value for the image label
+
+`ImageNetData` also has a method (`ReadFromCsv`) to read the image data `.csv` file and load it into an `ImageNetData` object.
+
+`ImageNetPipeline` is the class used for inspecting column values after the model has been trained. It has a `string` (`ImagePath`) for the image path. The `Label` is used to create and train the model, and it's also used to evaluate the model. The `PredictedLabelValue` is used during prediction and evaluation. For evaluation an input with training data, the predicted values, and the model are used.
```
Comment (luisquintanilla) [inline docs/machine-learning/tutorials/image-classification.md]: Rework second sentence. Too many ands. i.e. The `Label` is used to create, train and evaluate the model.

Rework fourth sentence. i.e. For evaluation, the inputs used are training data, predicted values and the model.

Comment (luisquintanilla) [inline docs/machine-learning/tutorials/image-classification.md]: Is `ImageNetPipeline` used in the code sample?

File: docs/machine-learning/tutorials/image-classification.md [hunk 25]
```diff
@@ -0,0 +1,400 @@
...
+Remove the existing class definition and add the following code, which has the `ImageNetPrediction` class, to the *ImageNetPrediction.cs* file:
+
+[!code-csharp[DeclareGlobalVariables](../../../samples/machine-learning/tutorials/TransferLearningTF/ImageNetPrediction.cs#DeclareTypes)]
+
+`ImageNetPrediction` is the image prediction class and has the following fields:
+
+* `Score` contains the confidence percentage for a given image classification.
+* `PredictedLabelValue` contains a value for the predicted image classification label.
+
+`ImageNetPrediction` is the class used for prediction after the model has been trained. It has a `string` (`ImagePath`) for the image path. The `Label` is used to create and train the model, and it's also used with a second dataset to evaluate the model. The `PredictedLabelValue` is used during prediction and evaluation. For evaluation, an input with training data, the predicted values, and the model are used.
```
Comment (luisquintanilla) [inline docs/machine-learning/tutorials/image-classification.md]: N/A

File: docs/machine-learning/tutorials/image-classification.md [hunk 26]
```diff
@@ -0,0 +1,400 @@
...
+
+ML.NET's transform pipelines compose a custom `transforms`set that is applied to your data before training or testing. The transforms' primary purpose is data [featurization](../resources/glossary.md#feature-engineering). Machine learning algorithms understand [featurized](../resources/glossary.md#feature) data, and when dealing with deep neural networks you must adapt the images to the format expected by the network. That format is a [numeric vector](../resources/glossary.md#numerical-feature-vector).
+
+When the model is trained and evaluated, by default, the values in the **Label** column are considered as correct values to be predicted. As you are using a pre-trained model you need to map fields to the new model. To do that, use the `MLContext.Transforms.Conversion.MapValueToKey`.  Name this `estimator` as you will then append the trainer to it. Add the next line of code:
+
+[!code-csharp[MapValueToKey1](../../../samples/machine-learning/tutorials/TransferLearningTF/Program.cs#MapValueToKey1)]
+
+Your image processing estimator uses pre-trained DNN featurizers for feature extraction. You use several image transforms to get the image data in the expected form:
```
Comment (luisquintanilla) [inline docs/machine-learning/tutorials/image-classification.md]: Write out DNN (Deep Neural Network) long form.

File: docs/machine-learning/tutorials/image-classification.md [hunk 27]
```diff
@@ -0,0 +1,400 @@
...
+
+1. The `LoadImages`transform images are loaded in memory as a Bitmap type. 
+2. The `Resize` transform resizes the images as the pre-trained model has a defined input image width and height.
+3. The `ImagePixelExtractingEstimator` transform extracts the pixels from the input images and converts them into a numeric vector.
+
+Add these image transforms as the next lines of code:
+
+[!code-csharp[ImageTransforms](../../../samples/machine-learning/tutorials/TransferLearningTF/Program.cs#ImageTransforms)]
+
+TensorFlow is a popular deep learning and machine learning toolkit that enables training deep neural networks (and general numeric computations), and is implemented as a transformer in ML.NET.  It's used to reuse the Inception model. The `TensorFlowTransform` extracts specified outputs (the model's image features `softmax2_pre_activation`) , and scores a dataset using a pre-trained Tensorflow model.
```
Comment (luisquintanilla) [inline docs/machine-learning/tutorials/image-classification.md]: I would move the first sentence (explaining what TensorFlow is further up in the tutorial). Then, in this section I'd mention that TensorFlow is implemented as a transformer in ML.NET.

Change last Tensorflow to TensorFlow

File: docs/machine-learning/tutorials/image-classification.md [hunk 28]
```diff
@@ -0,0 +1,400 @@
...
+
+[!code-csharp[ScoreTensorFlowModel](../../../samples/machine-learning/tutorials/TransferLearningTF/Program.cs#ScoreTensorFlowModel)]
+
+### Choose a learning algorithm
+
+To add the learning algorithm, call the `mlContext.MulticlassClassification.Trainers.LogisticRegression` wrapper method.  The `LogisticRegression` is appended to the `estimator` and accepts the Inception image features (`softmax2_pre_activation`) and the `Label` input parameters to learn from the historic data.  Add the trainer with the following code:
+
+[!code-csharp[AddTrainer](../../../samples/machine-learning/tutorials/TransferLearningTF/Program.cs#AddTrainer)]
+
+You also need to map the predictedlabel to the predictedlabelvalue:
```
Comment (luisquintanilla) [inline docs/machine-learning/tutorials/image-classification.md]: Emphasize / Inline-Code predictedlabel and/or predictedlabelvalue

File: docs/machine-learning/tutorials/image-classification.md [hunk 29]
```diff
@@ -0,0 +1,400 @@
...
+
+[!code-csharp[MapValueToKey2](../../../samples/machine-learning/tutorials/TransferLearningTF/Program.cs#MapValueToKey2)]
+
+The [Fit()](xref:Microsoft.ML.Trainers.MatrixFactorizationTrainer.Fit%28Microsoft.Data.DataView.IDataView,Microsoft.Data.DataView.IDataView%29) method trains your model with the provided training dataset. It executes the `Estimator` definitions by transforming the data and applying the training, and it returns back the trained model, which is a `Transformer`. The experiment is not executed until the `.Fit()` method runs. Fit the model to the `Train` data and return the trained model by adding the following as the next line of code in the `BuildAndTrainModel()` method:
```
Comment (luisquintanilla) [inline docs/machine-learning/tutorials/image-classification.md]: Consider using another word instead of experiment.

File: docs/machine-learning/tutorials/image-classification.md [hunk 30]
```diff
@@ -0,0 +1,400 @@
...
+## Classify images with a loaded model
+
+Add the following call to the `ClassifyImages` method as the next line of code in the `Main` method:
+
+[!code-csharp[CallClassifyImages](../../../samples/machine-learning/tutorials/TransferLearningTF/Program.cs#CallClassifyImages)]
+
+The `ClassifyImages` method executes the following tasks:
+
+* Loads the model.
+* Reads .csv file into List.
```
Comment (luisquintanilla) [inline docs/machine-learning/tutorials/image-classification.md]: Inline `List` or change to Reads .csv file into a list.

**Review by luisquintanilla — CHANGES_REQUESTED** [review]
Great writeup @JRAlexander . Made a few suggestions.

File: docs/machine-learning/tutorials/tensorflow-transfer-learning.md (hunk 8 above)
Comment (JRAlexander) [reply docs/machine-learning/tutorials/tensorflow-transfer-learning.md]: Refactored.

File: docs/machine-learning/tutorials/image-classification.md (hunk 24 above)
Comment (JRAlexander) [reply docs/machine-learning/tutorials/image-classification.md]: No, I am in process of revising this to match the latest version of the sample.

**Review by luisquintanilla — APPROVED** [review]
Thanks for the changes @JRAlexander. Looks good to me.
